toqito/state_opt/bell_inequality_max.py

```diff
@@ -0,0 +1,168 @@
+    r"""Compute the matrices M_a^x and N_b^y.
+
+    Args:
+        m: The number of measurement settings for Alice and Bob.
+        a: The specific measurement setting for Alice.
+        x: The specific measurement setting for Bob.
+
+    Returns:
+        The computed matrix.
+
+    """
+    # Create a permutation list as done in MATLAB
+    perm = list(range(m + 1))  # MATLAB indices start from 1
+    perm[0] = x
+    perm[x] = 0
+
+    # Calculate the matrix as in the MATLAB code
+    MN = a * np.eye(2 ** (m + 1)) + ((-1) ** a) * permutation_operator([2] * (m + 1), perm, 0, 1)
+
+    return MN
+
+
+def bell_inequality_max(joint_coe, a_coe, b_coe, a_val, b_val):
+    r"""Return the upper bound for the maximum violation(Tsirelson Bound) for a given bipartite Bell Inequality.
+
+    This computes the upper bound for the maximum value of a given bipartite Bell Inequality using SDP.
+    The method is from :cite:`Navascues_2014_Characterization` and the implementation is based on :cite:`QETLAB_link`.
+    This is useful for various tasks in device independent quantum information processing
+
+    The function formulates the problem as a SDP problem in the following format
+
+    .. math::
+        \[
+        \max \operatorname{tr} \left( W \cdot \sum_{a,b,x,y} B^{xy}_{ab} M^x_a \otimes N^y_b \right),
+        \]
+        \[
+        \text{s.t.} \quad \operatorname{tr}(W) = 1, \quad W \geq 0,
+        \]
+        \[
+        W^{T_P} \geq 0, \quad \text{for all bipartitions } P.
+        \]
+        \]
+
+    Examples
+    =======
+
+
+    Consider the I3322 Bell inequality
+    . . math::
+    I_{3322} = P(A_1 = B_1) + P(B_1 = A_2) + P(A_2 = B_2) + P(B_2 = A_3)
+           - P(A_1 = B_2) - P(A_2 = B_3) - P(A_3 = B_1) - P(A_3 = B_3)
+           \leq 2
+
+    The individual and joint coefficents and measurement values are encoded as matrices.
+    The upper bound can then be found in :code:'toqito' as follows.
+
+    >>> from toqito.state_opt import bell_inequality_max
+    >>> import numpy as np
+    >>> joint_coe = np.array([
+    ... [1, 1, -1],
+    ... [1, 1, 1],
+    ... [-1, 1, 0]
+    ... ])
+    >>> a_coe = np.array([0, -1, 0])
+    >>> b_coe = np.array([-1, -2, 0])
+    >>> a_val = np.array([0,1])
+    >>> b_val = np.array([0,1])
+    >>> '%.3f' % bell_inequality_max(joint_coe, a_coe, b_coe, a_val, b_val)
+    '0.250'
+
+    References
+    ==========
+    .. bibliography::
+        :filter: docname in docnames
+
+    :raises ValueError: If a_val or b_val are not length 2.
+    :param joint_coe: The coefficents for terms containing both A and B
+    :param a_coe: The coefficent for terms only containing A
+    :param b_coe: The coefficent for terms only containing B
+    :param a_val: The value of each measurement outcome for A
+    :param b_val: The value of each measurement outcome for B
+    :return: The upper bound for the maximum violation of the Bell inequality
+
+
+
+    """
+    m, _ = joint_coe.shape
+    oa = len(a_val)

```

Comment:
There's not really a reason to define these as separate variables as we can just check the length where necessary (see the comment below).